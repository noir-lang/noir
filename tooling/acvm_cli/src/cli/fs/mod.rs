pub(super) mod inputs;
pub(super) mod witness;
