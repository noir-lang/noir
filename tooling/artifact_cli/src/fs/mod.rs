pub mod artifact;
pub mod inputs;
pub mod witness;
