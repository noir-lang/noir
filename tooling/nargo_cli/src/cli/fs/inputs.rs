use noirc_abi::{
    input_parser::{Format, InputValue},
    Abi, InputMap, MAIN_RETURN_NAME,
};
use std::{collections::BTreeMap, path::Path};

use crate::errors::FilesystemError;

/// Returns the circuit's parameters and its return value, if one exists.
/// # Examples
///
/// ```ignore
/// let (input_map, return_value): (InputMap, Option<InputValue>) =
///   read_inputs_from_file(path, "Verifier", Format::Toml, &abi)?;
/// ```
pub(crate) fn read_inputs_from_file<P: AsRef<Path>>(
    path: P,
    file_name: &str,
    format: Format,
    abi: &Abi,
) -> Result<(InputMap, Option<InputValue>), FilesystemError> {
    if abi.is_empty() {
        return Ok((BTreeMap::new(), None));
    }

    let file_path = path.as_ref().join(file_name).with_extension(format.ext());
    if !file_path.exists() {
        return Err(FilesystemError::MissingTomlFile(file_name.to_owned(), file_path));
    }

    let input_string = std::fs::read_to_string(file_path).unwrap();
    let mut input_map = format.parse(&input_string, abi)?;
    let return_value = input_map.remove(MAIN_RETURN_NAME);

    Ok((input_map, return_value))
}
