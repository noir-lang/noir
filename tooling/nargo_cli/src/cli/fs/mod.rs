use std::{
    fs::File,
    io::Write,
    path::{Path, PathBuf},
};

pub(super) mod inputs;
pub(super) mod program;
pub(super) mod witness;

pub(super) fn create_named_dir(named_dir: &Path, name: &str) -> PathBuf {
    std::fs::create_dir_all(named_dir)
        .unwrap_or_else(|_| panic!("could not create the `{name}` directory"));

    PathBuf::from(named_dir)
}

pub(super) fn write_to_file(bytes: &[u8], path: &Path) -> String {
    let display = path.display();

    let mut file = match File::create(path) {
        Err(why) => panic!("couldn't create {display}: {why}"),
        Ok(file) => file,
    };

    match file.write_all(bytes) {
        Err(why) => panic!("couldn't write to {display}: {why}"),
        Ok(_) => display.to_string(),
    }
}
