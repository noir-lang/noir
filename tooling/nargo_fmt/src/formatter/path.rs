use noirc_frontend::{
    ast::{Path, PathKind, UnresolvedType},
    token::{Keyword, Token},
};

use super::Formatter;

impl<'a> Formatter<'a> {
    pub(super) fn format_path(&mut self, path: Path) {
        self.skip_comments_and_whitespace();

        match path.kind {
            PathKind::Plain => (),
            PathKind::Crate => {
                self.write_keyword(Keyword::Crate);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Dep => {
                self.write_keyword(Keyword::Dep);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Super => {
                self.write_keyword(Keyword::Super);
                self.write_token(Token::DoubleColon);
            }
        }

        for (index, segment) in path.segments.into_iter().enumerate() {
            if index > 0 {
                self.write_token(Token::DoubleColon);
            }
            self.write_identifier(segment.ident);

            if let Some(generics) = segment.generics {
                self.format_turbofish(generics);
            }
        }
    }

    pub(super) fn format_turbofish(&mut self, generics: Vec<UnresolvedType>) {
        self.write_token(Token::DoubleColon);
        self.write_token(Token::Less);
        for (index, typ) in generics.into_iter().enumerate() {
            if index > 0 {
                self.write_comma();
                self.write_space();
            }
            self.format_type(typ);
        }

        // Skip trailing comma, if any
        self.skip_comments_and_whitespace();
        if self.is_at(Token::Comma) {
            self.bump();
        }

        self.write_token(Token::Greater);
    }
}
