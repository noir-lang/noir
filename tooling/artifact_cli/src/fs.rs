use std::path::Path;

use color_eyre::eyre::{self, Context};
use noirc_artifacts::{contract::ContractArtifact, program::ProgramArtifact};

/// A parsed JSON build artifact.
#[derive(Debug, Clone)]
pub enum Artifact {
    Program(ProgramArtifact),
    Contract(ContractArtifact),
}

impl Artifact {
    /// Try to parse an artifact as a binary program or a contract
    pub fn read_from_file(path: &Path) -> eyre::Result<Artifact> {
        let json = std::fs::read(path)
            .with_context(|| format!("failed to read artifact from '{}'", path.display()))?;

        let as_program = || serde_json::from_slice::<ProgramArtifact>(&json).map(Artifact::Program);
        let as_contract =
            || serde_json::from_slice::<ContractArtifact>(&json).map(Artifact::Contract);

        Ok(as_program().or_else(|e| as_contract().map_err(|_| e))?)
    }
}
