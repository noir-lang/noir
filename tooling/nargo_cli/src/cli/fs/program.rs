use std::path::{Path, PathBuf};

use nargo::artifacts::{contract::ContractArtifact, program::ProgramArtifact};
use noirc_frontend::graph::CrateName;

use crate::errors::FilesystemError;

use super::{create_named_dir, write_to_file};

pub(crate) fn save_program_to_file<P: AsRef<Path>>(
    program_artifact: &ProgramArtifact,
    crate_name: &CrateName,
    circuit_dir: P,
) -> PathBuf {
    let circuit_name: String = crate_name.into();
    save_build_artifact_to_file(program_artifact, &circuit_name, circuit_dir)
}

pub(crate) fn save_contract_to_file<P: AsRef<Path>>(
    compiled_contract: &ContractArtifact,
    circuit_name: &str,
    circuit_dir: P,
) -> PathBuf {
    save_build_artifact_to_file(compiled_contract, circuit_name, circuit_dir)
}

fn save_build_artifact_to_file<P: AsRef<Path>, T: ?Sized + serde::Serialize>(
    build_artifact: &T,
    artifact_name: &str,
    circuit_dir: P,
) -> PathBuf {
    create_named_dir(circuit_dir.as_ref(), "target");
    let circuit_path = circuit_dir.as_ref().join(artifact_name).with_extension("json");

    write_to_file(&serde_json::to_vec(build_artifact).unwrap(), &circuit_path);

    circuit_path
}

pub(crate) fn read_program_from_file<P: AsRef<Path>>(
    circuit_path: P,
) -> Result<ProgramArtifact, FilesystemError> {
    let file_path = circuit_path.as_ref().with_extension("json");

    let input_string =
        std::fs::read(&file_path).map_err(|_| FilesystemError::PathNotValid(file_path))?;
    let program = serde_json::from_slice(&input_string)
        .map_err(|err| FilesystemError::ProgramSerializationError(err.to_string()))?;

    Ok(program)
}
