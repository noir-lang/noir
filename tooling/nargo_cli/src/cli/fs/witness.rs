use std::path::{Path, PathBuf};

use acvm::acir::native_types::WitnessStack;
use nargo::constants::WITNESS_EXT;

use super::{create_named_dir, write_to_file};
use crate::errors::FilesystemError;

pub(crate) fn save_witness_to_dir<P: AsRef<Path>>(
    witness_stack: WitnessStack,
    witness_name: &str,
    witness_dir: P,
) -> Result<PathBuf, FilesystemError> {
    create_named_dir(witness_dir.as_ref(), "witness");
    let witness_path = witness_dir.as_ref().join(witness_name).with_extension(WITNESS_EXT);

    let buf: Vec<u8> = witness_stack.try_into()?;

    write_to_file(buf.as_slice(), &witness_path);

    Ok(witness_path)
}
