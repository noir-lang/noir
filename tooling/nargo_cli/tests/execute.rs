#[allow(unused_imports)]
#[cfg(test)]
mod tests {
    // Some of these imports are consumed by the injected tests
    use assert_cmd::prelude::*;
    use predicates::prelude::*;

    use std::collections::BTreeMap;
    use std::fs;
    use std::path::PathBuf;
    use std::process::Command;

    use super::*;

    test_binary::build_test_binary_once!(mock_backend, "../backend_interface/test-binaries");

    // include tests generated by `build.rs`
    include!(concat!(env!("OUT_DIR"), "/execute.rs"));
}
