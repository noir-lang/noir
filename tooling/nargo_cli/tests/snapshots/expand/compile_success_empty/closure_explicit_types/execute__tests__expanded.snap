---
source: tooling/nargo_cli/tests/execute.rs
expression: expanded_code
---
fn ret_normal_lambda1() -> fn() -> Field {
    || -> Field 10
}

fn ret_closure1() -> fn[(Field,)]() -> Field {
    let x: Field = 20;
    || -> Field { x + 10 }
}

fn ret_closure2() -> fn[(Field, Field)]() -> Field {
    let x: Field = 20;
    let y: Field = 10;
    || -> Field { (x + y) + 10 }
}

fn ret_closure3() -> fn[(u32, u64)]() -> u64 {
    let x: u32 = 20;
    let y: u64 = 10;
    || -> u64 { ((x as u64) + y) + 10 }
}

fn accepts_closure1(f: fn[(Field,)]() -> Field) -> Field {
    f()
}

fn accepts_closure2(f: fn[(Field,)]() -> Field) -> fn[(Field,)]() -> Field {
    f
}

fn accepts_closure3(f: fn[(u32, u64)]() -> u64) -> u64 {
    f()
}

fn add_results<Env1, Env2>(f1: fn[Env1]() -> Field, f2: fn[Env2]() -> Field) -> Field {
    f1() + f2()
}

fn map<T, U, let N: u32, Env>(arr: [T; N], f: fn[Env](T) -> U) -> [U; N] {
    let first_elem: U = f(arr[0]);
    let mut ret: [U; N] = [first_elem; N];
    for i in 1..N {
        ret[i] = f(arr[i]);
    }
    ret
}

fn main() {
    assert(ret_normal_lambda1()() == 10);
    assert(ret_closure1()() == 30);
    assert(ret_closure2()() == 40);
    assert(ret_closure3()() == 40);
    let x: Field = 50;
    assert(accepts_closure1(|| -> Field x) == 50);
    assert(accepts_closure2(|| -> Field { x + 10 })() == 60);
    let y: u32 = 30;
    let z: u64 = 40;
    assert(accepts_closure3(|| -> u64 { (y as u64) + z }) == 70);
    let w: Field = 50;
    assert(add_results(|| -> Field 100, || -> Field x) == 150);
    assert(add_results(|| -> Field { x + 100 }, || -> Field { w + x }) == 250);
    let arr: [Field; 4] = [1, 2, 3, 4];
    assert(map(arr, |n: Field| -> Field { n + 1 }) == [2, 3, 4, 5]);
    assert(map(arr, |n: Field| -> Field { n + x }) == [51, 52, 53, 54]);
}
